### PR TITLE
Code review of all of banyan

### PR DESCRIPTION
**Pull request by rklaehn**
This is not going to be merged, just a way for people to comment on the code.

Everything below banyan-cli is just to exercise the code.

Most of the meat is in banyan/forest. banyan/tree.rs is a wrapper for a tree that peels off the empty case. banyan/query.rs is the querying stuff.

The rest (zstd_array.rs, thread_local_zstd.rs) is some compression plumbing that works but is not super pretty.

### CODE REVIEW

File: banyan/src/forest/mod.rs [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+        + DeserializeOwned
+        + Clone
+        + Debug
+        + FromIterator<Self::Key>
+        + Send
+        + Sync;
+    /// link type to use over block boundaries
+    type Link: Display + Debug + Hash + Eq + Clone + Copy + Send + Sync + DagCbor;
+
+    fn serialize_branch(links: Vec<(usize, Ipld)>, data: Vec<u8>) -> anyhow::Result<Vec<u8>>;
```
Comment (dvc94ch) [inline banyan/src/forest/mod.rs:58]: Why not `fn serialize_branch(links: BTreeMap<usize, Cid>, data: Vec<u8>) -> anyhow::Result<Vec<u8>>`?

Comment (rklaehn) [reply banyan/src/forest/mod.rs:58]: I want to have the option to have arbitary IPLD that is visible to the outside world. E.g. if we want to have some hints about the kind of data in the links.

Comment (dvc94ch) [reply banyan/src/forest/mod.rs:58]: and the trait itself makes sure you aren't tied to cbor?

Comment (dvc94ch) [reply banyan/src/forest/mod.rs:58]: I mean if you are just going to use cbor anyway why have `serialize_branch`, `deserialize_branch` and `Link` at all?

Comment (rklaehn) [reply banyan/src/forest/mod.rs:58]: So Link is mostly for the internal representation. A 32 byte sha2 hash is much smaller than a cid, even if both are `Copy`.

Comment (rklaehn) [reply banyan/src/forest/mod.rs:58]: Regarding the Ipld thing: I changed this to make it compatible with libipld. We can change it again as long as the on disk format stays the same.

**Comment by dvc94ch** [comment]
I guess we can close this at some point? :)

**Comment by rkuhn** [comment]
Yes :-)

-- 


*Actyx AG*, Dachauer Str. 192, 80992 Munich
HRB 223941 District Court 
Munich
Chairman of the Supervisory Board: Rolf Mathies
Executive Board: 
Oliver Stollmann (Chairman), Maximilian Fischer

**Pull request closed without merging**